cmd/publishing-bot/config/rules.go

```diff
@@ -172,12 +173,55 @@ func validateGoVersions(rules *RepositoryRules) (errs []error) {
 // go versions don't follow semver. Examples:
 // 1. 1.15.0 is invalid, 1.15 is valid
 // 2. 1.15.0-rc.1 is invalid, 1.15rc1 is valid
-var goVerRegex = regexp.MustCompile(`^([0-9]+)\.([0-9]\w*)(\.[1-9]\d*\w*)*$`)
+//
+// From go 1.21 onwards there is a change in the versioning format
+// Ref: https://tip.golang.org/doc/toolchain#versions
+// 1. 1.21 is invalid, and 1.21.0 is valid
+// 2. 1.21rc1 and 1.21.0rc1 are valid
+var goVerRegex = regexp.MustCompile(`^(?P<major>\d+)\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?(?:(?P<pre>alpha|beta|rc)\d+)?$`)
 
 func ensureValidGoVersion(version string) error {
-	if !goVerRegex.MatchString(version) {
+	match := goVerRegex.FindStringSubmatch(version)
+	if len(match) == 0 {
 		return fmt.Errorf("specified go version %s is invalid", version)
 	}
+
+	var majorVersion, minorVersion, patchVersion int
+	var preRelease string
+	patchVersionExists := false
+
+	majorVersion, err := strconv.Atoi(match[1])
+	if err != nil {
+		return fmt.Errorf("error parsing major version '%s' : %s", match[1], err)
+	}
+	minorVersion, err = strconv.Atoi(match[2])
+	if err != nil {
+		return fmt.Errorf("error parsing minor version '%s' : %s", match[2], err)
+	}
+	if match[3] != "" {
+		patchVersion, err = strconv.Atoi(match[3])
+		if err != nil {
+			return fmt.Errorf("error parsing patch version '%s' : %s", match[3], err)
+		}
+		patchVersionExists = true
+	}
+	preRelease = match[4]
+
+	// for go versions <= 1.20, patch version .0 should not exist
+	if majorVersion <= 1 && minorVersion <= 20 {
```

Comment:
Do we need the `majorVersion <= 1` check? 
Having a majorVersion is ensured by the regex itself iiuc. 
Thinking more -- I'm not sure if a go v2 would be there, so having the check would be fine if that is the intent.